docs/deployment/terraform-best-practices.md

```diff
@@ -0,0 +1,196 @@
+you may want to double-check your bucket and key values. If the bucket and key
+values are correct, then your state file is either _not_ marked as deleted or
+does not exist at all.
+
+Otherwise, you may remove the delete marker so that the state file no longer
+appears deleted. This will restore the previous version of the file and make it
+the latest version. Run the following command, using the same values for
+`BUCKET` and `KEY` as used in the previous command, and replacing `VERSION_ID`
+with the value output from the previous command:
+
+```bash
+aws s3api delete-object \
+  --bucket BUCKET \
+  --key KEY \
+  --version-id VERSION_ID
+```
+
+### Deny DeleteBucket Action
+
+As an additional measure to protect your Terraform state files from accidental
+loss, it is also recommended that you deny all users the ability to delete the
+bucket itself. At a later time, you may remove this protection when you are
+sure you want to delete the bucket.
+
+To perform this action this via the **AWS Management Console**:
+
+1. Go to the S3 service
+2. Go to the bucket used for storing state files
+3. Click the **Permissions** tab
+4. Click **Bucket Policy**
+5. Add the following policy statement to _deny_ the s3:DeleteBucket action for
+   all (`"*"`) principals, replacing `BUCKET_NAME` with the name of the bucket:
+
+   ```json
+   {
+     "Statement": [
+       {
+         "Sid": "DenyDeleteBucket",
+         "Effect": "Deny",
+         "Principal": "*",
+         "Action": "s3:DeleteBucket",
+         "Resource": "arn:aws:s3:::BUCKET_NAME"
+       }
+     ]
+   }
+   ```
+
+6. Click **Save**
+
+To perform this action via the **AWS CLI** instead, save the JSON shown above
+to a file named `policy.json` and run the following command from the directory
+in which you saved `policy.json`, replacing `BUCKET_NAME` with the name of the
+bucket:
+
+```bash
+aws s3api put-bucket-policy --bucket BUCKET_NAME --policy file://policy.json
+```
+
+Afterwards, remove the `policy.json` file.
+
+## Change Resources Only via Terraform
+
+**All resource changes must be made via Terraform**, otherwise you risk that
+your Terraform state file does not correctly represent the state of your
+deployment resources. Specifically, this means:
+
+- **DO NOT** change deployment resources via the AWS Management Console
+- **DO NOT** change deployment resources via the AWS CLI
+- **DO NOT** change deployment resources via any of the AWS SDKs
+
+Instead, change deployment resources **only** via changes to your Terraform
+files, along with subsequent Terraform commands.
+
+## Avoid Changing Connectivity Resources
+
+Don't mess with things that affect connectivity, like security groups
+
+## How to Destroy Everything
+
+### When Your State File Accurately Reflects Your Actual State
+
+How do I destroy all of my resources when I actually want to?
+
+### When Your State File Is Out of Sync with Your Actual State
```

Comment:
Given how open-ended this is, I'd be inclined to drop it until we have a more specific scenario.